File: backend/app/api/routers/notifications.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, select, update
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.platform import Notification
from app.models.user import User
from app.schemas.notification import NotificationList, NotificationOut

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=NotificationList)
def list_notifications(
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> NotificationList:
    """Recent in-app notifications for the current user + the unread count.
    Polled by the client (websockets are a deferred seam)."""
    unread = db.scalar(
        select(func.count(Notification.id)).where(
            Notification.user_id == user.id, Notification.seen.is_(False)
        )
    )
    items = db.scalars(
        select(Notification)
        .where(Notification.user_id == user.id)
        .order_by(Notification.created_at.desc())
        .limit(limit)
    ).all()
    return NotificationList(unread=unread or 0, items=[NotificationOut.model_validate(n) for n in items])


@router.post("/{notification_id}/seen", response_model=NotificationOut)
def mark_seen(
    notification_id: uuid.UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> NotificationOut:
    n = db.get(Notification, notification_id)
    if n is None or n.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found.")
    n.seen = True
    db.commit()
    db.refresh(n)
    return NotificationOut.model_validate(n)


@router.post("/read-all", status_code=status.HTTP_204_NO_CONTENT)
def mark_all_read(
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
) -> None:
    db.execute(
        update(Notification)
        .where(Notification.user_id == user.id, Notification.seen.is_(False))
        .values(seen=True)
    )
    db.commit()

File: frontend/src/api/notifications.ts
import { api } from "./client";

export type NotificationType =
  | "SALE"
  | "NEW_BID"
  | "BID_ACCEPTED"
  | "COLLAB_UPDATE"
  | "NEW_PRODUCTION_ORDER"
  | "SYSTEM";

export interface NotificationItem {
  id: string;
  type: NotificationType;
  title: string;
  body: string;
  link_url: string | null;
  seen: boolean;
  created_at: string;
}

export interface NotificationList {
  unread: number;
  items: NotificationItem[];
}

export async function getNotifications(limit = 20): Promise<NotificationList> {
  return (await api.get<NotificationList>(`/notifications?limit=${limit}`)).data;
}

export async function markNotificationSeen(id: string): Promise<NotificationItem> {
  return (await api.post<NotificationItem>(`/notifications/${id}/seen`)).data;
}

export async function markAllNotificationsRead(): Promise<void> {
  await api.post("/notifications/read-all");
}
